<a href="https://colab.research.google.com/github/siddharthprakash1/GoogleCollabNotebooks/blob/main/Question_and_Answer_Model_Using_Hugging_Face_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
context=("Pokémon[a] is a Japanese media franchise consisting of video games, animated series and films, a trading card game, and other related media. The franchise takes place in a shared universe in which humans co-exist with creatures known as Pokémon, a large variety of species endowed with special powers. The franchise's target audience is children aged 5 to 12,[1] but it is known to attract people of all ages.[2][3][4][5]"
"The franchise originated as a pair of role-playing games developed by Game Freak, from an original concept by its founder, Satoshi Tajiri. Released on the Game Boy on February 27, 1996, the games became sleeper hits and were followed by manga series, a trading card game, and anime series and films. From 1998 to 2000, Pokémon was exported to the rest of the world, creating an unprecedented global phenomenon dubbed Pokémania" "By 2002, the craze had ended, after which Pokémon became a fixture in popular culture, with new products being released to this day. In the summer of 2016, the franchise spawned a second craze with the release of Pokémon Go, an augmented reality game developed by Niantic. Pokémon has since been estimated to be the world's highest-grossing media franchise and one of the best-selling video game franchises."
"Pokémon has an uncommon ownership structure. Unlike most IPs, which are owned by one company, Pokémon is jointly owned by three: Nintendo, Game Freak, and Creatures.[6][7] Game Freak develops the core series role-playing games, which are published by Nintendo exclusively for their consoles, while Creatures manages the trading card game and related merchandise, occasionally developing spin-off titles." "The three companies established The Pokémon Company (TPC) in 1998 to manage the Pokémon property within Asia. The Pokémon anime series and films are co-owned by Shogakukan. Since 2009, The Pokémon Company International (TPCi) subsidiary of TPC has managed the franchise in all regions outside of Asia.[8]"
"Name"
"The original full name of the franchise is Pocket Monsters (ポケットモンスター, Poketto Monsutā), which was abbreviated to Pokemon during development of the original games. When the franchise was released internationally, the short form of the title was used, with an acute accent (´) over the e to aid in pronunciation.[9]"
"Pokémon refers to both the franchise itself and the creatures within its fictional universe. As a noun, it is identical in both the singular and plural, as is every individual species name;[10] it is grammatically correct to say 'one Pokémon' and 'many Pokémon', as well as 'one Pikachu' and 'many Pikachu'. In English, Pokémon may be pronounced either /'powkɛmon/ (poe-keh-mon) or /'powkɪmon/ (poe-key-mon).[11]"
"General concept"
"Main article: Gameplay of Pokémon"
"The Pokémon franchise is set in a world in which humans coexist with creatures known as Pokémon. Pokémon Red and Blue contain 151 Pokémon species, with new ones being introduced in subsequent games; as of December 2023, 1,025 Pokémon species have been introduced.[b] Most Pokémon are inspired by real-world animals;[12] for example, Pikachu are a yellow mouse-like species[13] with lightning bolt-shaped tails[14] that possess electrical abilities.[15]"
"The player character takes the role of a Pokémon Trainer. The Trainer has three primary goals: travel and explore the Pokémon world; discover and catch each Pokémon species in order to complete their Pokédex; and train a team of up to six Pokémon at a time and have them engage in battles. Most Pokémon can be caught with spherical devices known as Poké Balls. Once the opposing Pokémon is sufficiently weakened, the Trainer throws the Poké Ball against the Pokémon, which is then transformed into a form of energy and transported into the device. Once the catch is successful, the Pokémon is tamed and is under the Trainer's command from then on. If the Poké Ball is thrown again, the Pokémon re-materializes into its original state. The Trainer's Pokémon can engage in battles against opposing Pokémon, including those in the wild or owned by other Trainers. Because the franchise is aimed at children, these battles are never presented as overtly violent and contain no blood or gore.[I] Pokémon never die in battle, instead fainting upon being defeated.[20][21][22]"
"After a Pokémon wins a battle, it gains experience and becomes stronger.[23] After gaining a certain amount of experience points, its level increases, as well as one or more of its statistics. As its level increases, the Pokémon can learn new offensive and defensive moves to use in battle.[24][25] Furthermore, many species can undergo a form of spontaneous metamorphosis called Pokémon evolution, and transform into stronger forms.[26] Most Pokémon will evolve at a certain level, while others evolve through different means, such as exposure to a certain item.[27]"
"History"
"For a chronological guide, see Timeline of Pokémon."
"Origins"
"The main idea behind Pokémon was conceived by Satoshi Tajiri. Tajiri grew up in Machida, a suburb of Tokyo. In his youth, he enjoyed discovering and catching insects and other small creatures in the various ponds and fields that surrounded his town.[28][29] As Japan's economic miracle occurred, many cities, including Machida, were significantly expanded. As a consequence, Machida's nature was largely destroyed. In his second year of junior high school,[30][31] an arcade hall opened in Tajiri's neighborhood, introducing him to video games. While studying electrical engineering at Tokyo College of Technology [ja], Tajiri began publishing a doujinshi magazine titled Game Freak.[32] The title was inspired by the 1932 film Freaks, which Tajiri was fascinated with at the time.[33] He self-published the first issue of the magazine in March 1983, at the age of 17.[34][35] At the time, magazines specializing in video games did not yet exist in Japan, allowing Game Freak to fill a gap in the market.[36][37] Concurrently, Tajiri was contacted by aspiring manga artist Ken Sugimori, who became Game Freak's illustrator.[38] Game Freak folded in the late 1980s,[34] by which point Tajiri had become a respected game journalist in Japan's fledgling video game industry.[39]"
"Tajiri's knowledge of video games brought him into contact with Tsunekazu Ishihara. Ishihara had studied Arts and Science at Tsukuba University, and was trained in CGI. After working in advertising for two years, Ishihara joined a company named Sedic in 1983, which created video graphics and software,[40] including one video game, Otocky (1987).[41][42] As video games grew in popularity, Sedic also produced a number of game-related television shows for Fuji Television's late-night slot.[40] In his job as a TV producer, Ishihara befriended Tajiri, as well as Shigesato Itoi, who would later become CEO of Ape, Inc.[43][44] Ishihara was also the general director of the world's first gaming encyclopedia: TV Games: Encyclopedia of Video Games (テレビゲーム―電視遊戯大全, Terebigēmu ― Denshi Yūgi Taizen).[45][46] Hatakeyama & Kubo (2000) wrote that, at the time, Ishihara was probably the most well-connected man in the industry,[47] with a knowledge of games 'beyond comparison'.[48] Ishihara was also interested in playing cards,[49][50] and contributed to the development of at least three simple card games designed by Itoi for Ape.[c]"
)
context

"Pokémon[a] is a Japanese media franchise consisting of video games, animated series and films, a trading card game, and other related media. The franchise takes place in a shared universe in which humans co-exist with creatures known as Pokémon, a large variety of species endowed with special powers. The franchise's target audience is children aged 5 to 12,[1] but it is known to attract people of all ages.[2][3][4][5]The franchise originated as a pair of role-playing games developed by Game Freak, from an original concept by its founder, Satoshi Tajiri. Released on the Game Boy on February 27, 1996, the games became sleeper hits and were followed by manga series, a trading card game, and anime series and films. From 1998 to 2000, Pokémon was exported to the rest of the world, creating an unprecedented global phenomenon dubbed PokémaniaBy 2002, the craze had ended, after which Pokémon became a fixture in popular culture, with new products being released to this day. In the summer of 20

In [4]:
questions=[
    "what is Pokémon?","what is the target audience of Pokémon?","who is the founder of Pokémon?","when was pokemon a big hit?","Where did the founder of pokemon stay?"
]

In [5]:
from transformers import BertForQuestionAnswering

this is the pytorch implementation of bert if u want tesnorflow then add tf

## Importing BertForQuestionAnswering

This line imports the `BertForQuestionAnswering` class from the `transformers` library. This class provides a pre-trained model specifically designed for question-answering tasks.

The `transformers` library offers various pre-trained models for natural language processing (NLP) tasks, including question answering. `BertForQuestionAnswering` is a fine-tuned version of the Bidirectional Encoder Representations from Transformers (BERT) model, specifically for extracting answers from a given context based on a posed question.

By importing this class, you gain access to a powerful tool for building your question-answering model without the need to train a model from scratch. The pre-trained model has already learned valuable representations of words and their relationships, which can be leveraged for your task.


In [6]:
!pip install huggingface-cli

In [ ]:
!huggingface-cli login


    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

    To login, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Token: 
Add token as git credential? (Y/n) y
Token is valid (permission: read).
Cannot authenticate through git-credential as no helper is defined on your machine.
You might have to re-authenticate when pushing to the Hugging Face Hub.
Run the following command in your terminal in case you want to set the 'store' credent

In [7]:
model=BertForQuestionAnswering.from_pretrained('deepset/bert-base-cased-squad2')

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/508 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/433M [00:00<?, ?B/s]

Some weights of the model checkpoint at deepset/bert-base-cased-squad2 were not used when initializing BertForQuestionAnswering: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


## Loading the Pre-trained Question Answering Model

This line creates a question-answering model instance using the `BertForQuestionAnswering.from_pretrained()` method. Let's break down what's happening:

* **`model =`**: Assigns the created model to the variable `model`.
* **`BertForQuestionAnswering`**: This refers to the class imported earlier from the `transformers` library. It provides functionalities for question-answering tasks.
* **`.from_pretrained('deepset/bert-base-cased-squad2')`**: This part performs the following actions:
    * **`from_pretrained()`**: This method is used to load a pre-trained model from the Hugging Face model hub.
    * **`'deepset/bert-base-cased-squad2'`**: This is the specific model identifier you're using. It points to a pre-trained BERT model from the Hugging Face model hub:
        * **`deepset/`**: This indicates the model is provided by a contributor named "deepset".
        * **`bert-base-cased-squad2`**: This specifies the model architecture (BERT), case sensitivity (cased), and fine-tuning (on the SQuAD 2 dataset for question answering).

In essence, this line loads a pre-trained model that has already been trained on a massive dataset of text and question-answer pairs. This pre-trained knowledge can be effectively applied to your specific question-answering task without the need to train a model from scratch.


In [8]:
from transformers import AutoTokenizer

## Importing AutoTokenizer

This line imports the `AutoTokenizer` class from the `transformers` library. This class plays a crucial role in working with pre-trained models from the Hugging Face model hub.

**Functionality of AutoTokenizer:**

The `AutoTokenizer` simplifies the process of loading the appropriate tokenizer for a given pre-trained model. It automatically determines the correct tokenizer based on the model identifier you provide later in your code.

**Benefits of AutoTokenizer:**

* **Convenience:** Saves you from manually searching for the corresponding tokenizer for each model.
* **Efficiency:** Ensures compatibility between the model and tokenizer.
* **Flexibility:** Works with various pre-trained models from the Hugging Face model hub.

In essence, `AutoTokenizer` streamlines the process of preparing text data for use with your pre-trained model by handling the underlying complexities of tokenizer selection.


In [9]:
tokenizer=AutoTokenizer.from_pretrained('deepset/bert-base-cased-squad2')

tokenizer_config.json:   0%|          | 0.00/152 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

so here bert understands string using tokens so we  have to tokenize the string input to pass it as an input in the bert model

## Loading the Pre-trained Tokenizer

This line creates a tokenizer instance using the `AutoTokenizer.from_pretrained()` method. Let's break down the components:

* **`tokenizer =`**: Assigns the created tokenizer to the variable `tokenizer`.
* **`AutoTokenizer`**: This refers to the class imported earlier from the `transformers` library. It provides functionalities for preparing text data for use with pre-trained models.
* **`.from_pretrained('deepset/bert-base-cased-squad2')`**: This part performs the following actions:
    * **`from_pretrained()`**: This method is used to load a tokenizer that is compatible with a pre-trained model from the Hugging Face model hub.
    * **`'deepset/bert-base-cased-squad2'`**: This is the same model identifier you used earlier to load the question-answering model (`BertForQuestionAnswering`). By using the same identifier here, you ensure compatibility between the tokenizer and the model.



In [10]:
tokenizer.encode(questions[0],truncation=True,padding=True)

[101, 1184, 1110, 22926, 136, 102]

101 here refers to the special token of bert ['cls'] and 102 for the end

so here padding aand truncation both are true coz when using bert it wants 512 tokens as input but the question barely has 10-15 tokens so we need to add padding and similarly if the question is too big we need to truncate

"[cls]<context>[sep]<question>[sep][pad][pad]" this is how the question is gonna go as the input in the bert model

## Encoding the Question with Tokenizer

This line uses the `tokenizer.encode()` method to prepare the first question (stored in the `questions[0]` variable) for the pre-trained model. Here's a breakdown:

* **`tokenizer.encode()`**: This method is a core functionality of the tokenizer you loaded earlier. It takes text data as input and performs the following actions:
    * **Word Segmentation:** Breaks down the question into tokens (words or sub-word units) based on the model's vocabulary.
    * **Adding Special Tokens:** May add special tokens (e.g., start/end of sentence markers) if required by the model.
    * **Numerical Representation:** Converts each token into a numerical value (id) based on the model's vocabulary.
    * **Output:** Returns a list of these numerical token ids representing the question.
* **`questions[0]`**: This refers to the first element in a list named `questions`. It's assumed this list


In [11]:
from transformers import pipeline

In [12]:
nlp=pipeline('question-answering',model=model,tokenizer=tokenizer)

## Creating a Question-Answering Pipeline

This line creates a question-answering pipeline using the `pipeline()` function from the `transformers` library. This pipeline simplifies the process of using your pre-trained model for question-answering tasks.

* **`nlp =`**: Assigns the created pipeline to the variable `nlp`.
* **`pipeline('question-answering', ...)`**: This part creates the pipeline.
    * **`'question-answering'`**: This argument specifies the type of pipeline to create. In this case, you're building a pipeline for question-answering tasks.
    * **`model=model`**: This argument assigns the pre-trained question-answering model (`model`) you loaded earlier to the pipeline.
    * **`tokenizer=tokenizer`**: This argument assigns the tokenizer (`tokenizer`) you loaded earlier to the pipeline.

**Benefits of Using a Pipeline:**

The pipeline provides a convenient and high-level interface for interacting with your question-answering model. It abstracts away the underlying complexities of model inputs and outputs, allowing you to focus on feeding questions and retrieving answers.





In [13]:
nlp({
    'question':questions[0],
    'context': context
})

{'score': 0.268502414226532,
 'start': 1254,
 'end': 1297,
 'answer': 'Pokémon has an uncommon ownership structure'}

In [14]:
nlp({
    'question':questions[1],
    'context': context
})

{'score': 0.9904259443283081, 'start': 4086, 'end': 4094, 'answer': 'children'}

In [15]:
nlp({
    'question':questions[2],
    'context': context
})

{'score': 0.9954031109809875,
 'start': 544,
 'end': 558,
 'answer': 'Satoshi Tajiri'}

In [16]:
nlp({
    'question':questions[3],
    'context': context
})

{'score': 0.19736699759960175, 'start': 601, 'end': 605, 'answer': '1996'}

In [17]:
nlp({
    'question':questions[4],
    'context': context
})

{'score': 0.0005179457366466522,
 'start': 5382,
 'end': 5403,
 'answer': "Tajiri's neighborhood"}